### PR TITLE
02 train elasticnet

### PR DESCRIPTION
**Pull request by bill-amadio**
####Purpose/Implementation

This script is the second in a planned three-script workflow to predict reported_gender from expression data.

_What scientific question is your analysis addressing? 
What was your approach?
If this is not adding an analysis, describe your changes in this section._

This script accepts command line arguments to locate and read training and target sets generated by script 01-clean_split_data.R.  An additional command line argument specifies the median absolute deviation filtering threshold for transcripts in the training set.

A logistic regression predictive model is trained using glmnet and glmnetUtils.  To support model evaluation in the next script, the model object for the best fitting combination of lambda and alpha is saved along with the indexes of the filtered transcripts.

A .tsv file of non-zero coefficients and transcripts is also saved.

####Issue

_What GitHub issue does your pull request address?_

84: RNA-based Prediction of Sex

####Direction for reviewers

_Tell potential reviewers what kind of feedback you are soliciting. 
Are there particular areas that need a closer look?
Is there something you want to discuss further?
Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?_

I only discovered the need to save the filtered transcript indexes when I tried to go ahead to the work of script 3 (evaluation) in the sequence.  The filtered transcripts can certainly be reconstructed by reading the training set again, so I'd like to hear your preferences on bare bones saving/reconstruction v. detailed saving anticipating future usage of the saved items.

Do we want to hold off merging this script into the master until the functionality of the evaluation script is complete?

#### Results

_If your pull request includes code that produces scientific results, please summarize the results here.
This can help facilitate discussion around interpretation.
Please state what kinds of results are included (e.g., table, figure)._

Models filtering the top 5% and top 25% median absolute deviation transcripts both achieved 98% accuracy on a manual check of the test set.  The 5% model used 23 features, and the 25% model used 41 features.  Will there be any biological followup of the model features or any interest in optimizing training parameters and choices?

#### Docker and continuous integration

_Check all those that apply or remove this section if it is not applicable._

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

### CODE REVIEW

**Comment by jaclyn-taroni** [comment]
Hi @bill-amadio, thank you for this! I'm planning on taking a look today. I will now update your branch and make changes to the CI config file to fix the build error that we have currently + add this new script as well.

**Comment by jaclyn-taroni** [comment]
> I only discovered the need to save the filtered transcript indexes when I tried to go ahead to the work of script 3 (evaluation) in the sequence. The filtered transcripts can certainly be reconstructed by reading the training set again, so I'd like to hear your preferences on bare bones saving/reconstruction v. detailed saving anticipating future usage of the saved items.

Hmm... when you asked about what step to perform the MAD filtering in the in `#open-pbta` channel in slack, I thought this step was the right spot for it. But if you moved it to the `01-clean_split_data.R`, would you have to save the filtered transcripts? You could calculate the MAD only on the training set and use the indices to filter both the training set and test set, saving the filtered matrices. What do you think? 

If this seems like a change you might want to make, it doesn't have to happen in this pull request. We can re-visit once the evaluation script goes through.

> Do we want to hold off merging this script into the master until the functionality of the evaluation script is complete?

We can merge this to master first and if anything comes up when you are working on the evaluation, you can make changes to this script in that new `03`-related branch. If a pull request stays open for too long and pull requests from other folks get merged, it creates more opportunities for merge conflicts, etc. to arise. 

> Will there be any biological followup of the model features or any interest in optimizing training parameters and choices?

I think we want to prioritize the evaluations on `germline_sex_estimate` and see how that goes. (Here I'm assuming your 98% accuracy was on the reported gender.) That will inform next steps. Essentially, if model performs well on the `germline_sex_estimate`, then biological followup is more warranted.

**Comment by bill-amadio** [comment]
Pull seemed to work fine. I replaced the print/quits with stops, and pushed the changes.  

Merge if you are satisfied and the decision on MAD filtering can wait.

**Comment by bill-amadio** [comment]
Regarding the placement of MAD filtering.  Your suggested workflow certainly will work.  Can you give me an idea of how the scripts will be used?  Will someone ever want to try several filtering thresholds?  And if so, will the multiple training sets become difficult to navigate?

Do we want to save the full training and test sets regardless of the choice of filtering threshold?

Will these scripts go into some kind of production platform?  Will folks be using them going forward for different purposes?  

Thanks.

**Comment by jaclyn-taroni** [comment]
> Can you give me an idea of how the scripts will be used? Will someone ever want to try several filtering thresholds? And if so, will the multiple training sets become difficult to navigate?

I think when I wrote the ideas behind the structure of this on https://github.com/AlexsLemonade/OpenPBTA-analysis/pull/123#issuecomment-540150540 I assumed someone might want to try several thresholds (and you yourself have tried different thresholds!). If that's the case, then moving MAD into `01` would probably make things a bit difficult to work with as you suggest. I am less enthusiastic about that move, now, and therefore...

> Do we want to save the full training and test sets regardless of the choice of filtering threshold?

I think we do want to do that in `01` and keep the MAD filtering in `02`.

> Will these scripts go into some kind of production platform? Will folks be using them going forward for different purposes?

The general use case, e.g., predicting germline sex from RNA-seq data, is likely to come up again for folks in different settings. That means that things like what columns the IDs and labels are in will change in those different settings.

**Comment by jaclyn-taroni** [comment]
Thank you @bill-amadio! Merging now.
